drivers/sensor/si7006/si7006.c

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2019 Electronut Labs
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <sensor.h>
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <string.h>
+#include <sys/byteorder.h>
+#include <sys/__assert.h>
+#include <logging/log.h>
+#include <i2c.h>
+#include <logging/log.h>
+#include <stdio.h>
+#include "si7006.h"
+
+#define LOG_LEVEL CONFIG_SENSOR_LOG_LEVEL
+LOG_MODULE_REGISTER(si7006);
+
+struct si7006_data {
+	struct device *i2c_dev;
+	u16_t temperature;
+	u16_t humidity;
+};
+
+/**
+ * @brief function to get relative humidity
+ *
+ * @return int 0 on success
+ */
+static int si7006_get_humidity(struct device *i2c_dev,
+	struct si7006_data *si_data)
+{
+	int retval = 0;
+	uint8_t hum[2];
+
+	if (i2c_burst_read(i2c_dev, DT_INST_0_SILABS_SI7006_BASE_ADDRESS,
+			SI7006_MEAS_REL_HUMIDITY_MASTER_MODE, hum, sizeof(hum))
+			!= 0) {
+		LOG_ERR("read register err");
+		retval = -1;
+	}
+
+	si_data->humidity = (hum[0] << 8) | (hum[1]);
+
+	return retval;
+}
+
+/**
+ * @brief function to get temperature
+ *
+ * @return int 0 on success
+ */
+
+static int si7006_get_temperature(struct device *i2c_dev,
+	struct si7006_data *si_data)
+{
+	uint8_t temp[2];
+	int retval = 0;
+
+	if (i2c_burst_read(i2c_dev, DT_INST_0_SILABS_SI7006_BASE_ADDRESS,
+			SI7006_MEAS_TEMP_MASTER_MODE, temp, sizeof(temp))
+			!= 0){
+		LOG_ERR("read register err");
+		retval = -1;
+	}
+
+	si_data->temperature = (temp[0] << 8) | (temp[1]);
+
+	return retval;
+}
+
+/**
+ * @brief fetch a sample from the sensor
+ *
+ * @return 0
+ */
+static int si7006_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct si7006_data *si_data = dev->driver_data;
+
+	si7006_get_temperature(si_data->i2c_dev, si_data);
+	si7006_get_humidity(si_data->i2c_dev, si_data);
+
+	return 0;
+}
+
+/**
+ * @brief sensor value get
+ *
+ * @return -ENOTSUP for unsupported channels
+ */
+static int si7006_channel_get(struct device *dev, enum sensor_channel chan,
+				  struct sensor_value *val)
+{
+	struct si7006_data *si_data = dev->driver_data;
+
+	if (chan == SENSOR_CHAN_AMBIENT_TEMP) {
+
+		s64_t temp_cal = (((17572 * si_data->temperature) >> 16) - 4685)
```

Comment:
The variable name is confusing: `_cal` suggests it's a calibrated temperature.  It's actually in micro-Celsius, so `_uCel` [would be ideal](https://unitsofmeasure.org/ucum.html#section-Prefixes) except that apparently that violates a strict rule against camelcase.  @carlescufi maybe chime in here: that rule is *not* helping maintainability and code clarity.

The inner expression should be `((17572 * (s32_t)si_data->temperature) / 65536)` for a couple reasons.  You're implicitly converting a `u16_t` to a signed value in a context where a larger integral type is needed: make that explicit.  Shifting negative signed values is implementation-defined; division is well-defined.  The compiler will do the optimal thing.

Finally `s32_t` is sufficient to hold the result value.

Same for RH below.